src/components/shared/MarkerContextMenu.js

```diff
@@ -224,20 +240,57 @@ class MarkerContextMenuImpl extends PureComponent<Props> {
   };
 
   render() {
-    const { marker } = this.props;
+    const { marker, previewSelection, committedRange } = this.props;
     const { data } = marker;
+
+    const selectionEnd = previewSelection.hasSelection
+      ? previewSelection.selectionEnd
+      : committedRange.end;
+
+    const markerEnd = marker.end || marker.start;
+
     return (
       <ContextMenu
         id="MarkerContextMenu"
         onShow={this._onShow}
         onHide={this._onHide}
       >
-        <MenuItem onClick={this.setStartRange}>
-          Set selection start time here
-        </MenuItem>
-        <MenuItem onClick={this.setEndRange}>
-          Set selection end time here
-        </MenuItem>
+        {this._isZeroDurationMarker(marker) ? (
+          <>
+            <MenuItem onClick={this.setStartRangeFromMarkerStart}>
+              Set selection start time here
+            </MenuItem>
+            <MenuItem onClick={this.setEndRangeFromMarkerEnd}>
+              Set selection end time here
+            </MenuItem>
+          </>
+        ) : (
+          <>
+            <SubMenu title="Set selection start">
+              <MenuItem onClick={this.setStartRangeFromMarkerStart}>
+                From the start of this marker
+              </MenuItem>
+              <MenuItem
+                onClick={this.setStartRangeFromMarkerEnd}
+                disabled={markerEnd > selectionEnd}
+              >
+                From the end of this marker
+              </MenuItem>
+            </SubMenu>
+            <SubMenu title="Set selection end">
+              <MenuItem onClick={this.setEndRangeFromMarkerStart}>
+                From the start of this marker
+              </MenuItem>
+              <MenuItem
+                onClick={this.setEndRangeFromMarkerEnd}
+                disabled={markerEnd > selectionEnd}
+              >
+                From the end of this marker
+              </MenuItem>
+            </SubMenu>
```

Comment:
The `disabled` property should be put on the other menuitem (`setEndRangeFromMarkerStart`), because this is the one that can show a problem. `setEndRangeFromMarkerEnd` shouldn't show a problem of the end being before the start. Indeed, because this marker is currently displayed, it means that its end is after the current start. It can be confusing thinking about it, but it's a lot easier if you draw these things on a piece of paper.

So, please add the `disabled` attribute on the other `MenuItem`, and remove it here.
__The condition needs to be changed too__ but I want that you think about this so I won't tell you. **Again** it's a lot easier to think about this with a pen and a sheet of paper. Even myself, I do it like this.

Reply:
This is what I come up with

In this, if `marker.start` is before the `selectionStart`
we shouldn't set `selectionEnd` to `marker.start` as that only creates the problem  of getting `selectionEnd` before `selectionStart`
```suggestion
            <SubMenu title="Set selection end">
              <MenuItem onClick={this.setEndRangeFromMarkerStart}
                disabled={selectionStart > marker.start}
              >
                From the start of this marker
              </MenuItem>
              <MenuItem
                onClick={this.setEndRangeFromMarkerEnd}
              >
                From the end of this marker
              </MenuItem>
            </SubMenu>
```

Reply:
This looks reasonable!